lightningd/invoice.c

```diff
@@ -1808,3 +1808,92 @@ static const struct json_command createinvoice_command = {
 };
 
 AUTODATA(json_command, &createinvoice_command);
+
+static struct command_result *json_preapproveinvoice(struct command *cmd,
+						     const char *buffer,
+						     const jsmntok_t *obj UNNEEDED,
+						     const jsmntok_t *params)
+{
+	const char *invstring;
+	struct json_stream *response;
+	bool approved;
+
+	if (!param(cmd, buffer, params,
+		   /* FIXME: parameter should be invstring now */
+		   p_req("bolt11", param_string, &invstring),
+		   NULL))
+		return command_param_failed();
+
+	/* Strip optional URI preamble. */
+	if (strncmp(invstring, "lightning:", 10) == 0 ||
+	    strncmp(invstring, "LIGHTNING:", 10) == 0)
+		invstring += 10;
+
+	u8 *msg = towire_hsmd_preapprove_invoice(NULL, invstring);
+
+	if (!wire_sync_write(cmd->ld->hsm_fd, take(msg)))
```

Comment:
Doing synchronous comms here may not be what we want. It ends up blocking the entire `lightningd` until the signer has responded. This means that if we have a long RTT from `lightningd` to `hsmd` we are unresponsive for that roundtrip. Worse, if the signer does not respond we are blocked forever and can't react to on-chain events and peer messsages.

We should always prefer async comms in `lightningd`.